qgym/benchmarks/metrics/routing_metrics.py

```diff
@@ -0,0 +1,206 @@
+                interaction_fidelities[index] *= connection_graph[qubit1][qubit2][
+                    "weight"
+                ]
+            else:
+                interactions.append(interaction)
+                interaction_fidelities.append(
+                    connection_graph[qubit1][qubit2]["weight"]
+                )
+
+        final_fidelity_mean = mean(interaction_fidelities)
+
+        return initial_fidelity_mean / final_fidelity_mean
+
+    def average_qubit_fidelity_ratio_loss(
+        self,
+        initial_interaction_circuit: NDArray[np.int_],
+        connection_graph: nx.Graph,
+        final_interaction_circuit: NDArray[np.int_],
+    ) -> int:
+        """Method which calculates the average fidelity on qubits, where the fidelity of
+        one qubit is calculated by consecutively multiplying 1 with the
+        connection_graph_edge_fidelity for those edge in the interaction graph that are
+        connected to that particular qubit. The averages over all qubits before and
+        after adding swaps are calculated and divided to get the ratio.
+
+        Args:
+            initial_interaction_circuit: An array of 2-tuples of integers, where every
+                tuple represents a, not specified, gate acting on the two qubits labeled
+                by the integers in the tuples.
+            connection_graph: :class:`networkx.Graph` representation of the QPU
+                topology. Each node represents a physical qubit and each edge represents
+                a connection in the QPU topology.
+            final_interaction_circuit: An array of 2-tuples of integers, where every
+                tuple represents a, not specified, gate acting on the two qubits labeled
+                by the integers in the tuples.
+        """
+
+        initial_interaction_qubits = []
+        for qubit1, qubit2 in initial_interaction_circuit:
+            if qubit1 not in initial_interaction_qubits:
+                initial_interaction_qubits.append(qubit1)
+            if qubit2 not in initial_interaction_qubits:
+                initial_interaction_qubits.append(qubit2)
+
+        initial_qubit_fidelities = []
+        for qubit in initial_interaction_qubits:
+            qubit_fidelity = 1
+            for qubit1, qubit2 in initial_interaction_circuit:
+                if qubit == qubit1 or qubit == qubit2:
+                    qubit_fidelity *= connection_graph[qubit1][qubit2]["weight"]
+            initial_qubit_fidelities.append(qubit_fidelity)
+        initial_qubit_fidelity_mean = mean(initial_qubit_fidelities)
+
+        final_qubit_fidelities = []
+        for qubit in initial_interaction_qubits:
+            qubit_fidelity = 1
+            for qubit1, qubit2 in final_interaction_circuit:
+                if qubit == qubit1 or qubit == qubit2:
+                    qubit_fidelity *= connection_graph[qubit1][qubit2]["weight"]
+            final_qubit_fidelities.append(qubit_fidelity)
```

Comment:
Same as above.

Even the code can be reduce to a one-liner, I would probably put it into a meaningful function, e.g., `calculate_fidelities`, so this function can be reduced to a few calls, like 5 or 6 lines of code.